File: server/src/test/java/com/fluffy/reaction/domain/LikeServiceTest.java
package com.fluffy.reaction.domain;

import com.fluffy.auth.domain.Member;
import com.fluffy.auth.domain.MemberRepository;
import com.fluffy.support.AbstractIntegrationTest;
import com.fluffy.support.data.MemberTestData;
import static org.assertj.core.api.Assertions.assertThat;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

class LikeServiceTest extends AbstractIntegrationTest {

    @Autowired
    private LikeService likeService;

    @Autowired
    private ReactionRepository reactionRepository;

    @Autowired
    private MemberRepository memberRepository;

    @Test
    @DisplayName("좋아요를 할 수 있다.")
    void like() {
        // given
        Member member = createMember();
        Reaction reaction = new Reaction("EXAM", 1L, member.getId(), ReactionType.LIKE);
        reactionRepository.save(reaction);

        // when
        likeService.like(new Like(LikeTarget.EXAM, 1L), member.getId());

        // then
        Reaction savedReaction = reactionRepository.findById(reaction.getId()).orElseThrow();
        assertThat(savedReaction.getStatus()).isEqualTo(ReactionStatus.ACTIVE);
    }

    @Test
    @DisplayName("이전에 좋아요를 취소한 상태에서는 좋아요를 새로 생성하지 않고, 기존 좋아요를 활성화한다.")
    void like_alreadyUnliked() {
        // given
        Member member = createMember();
        Reaction reaction = new Reaction("EXAM", 1L, member.getId(), ReactionType.LIKE);
        reaction.delete();
        reactionRepository.save(reaction);

        // when
        likeService.like(new Like(LikeTarget.EXAM, 1L), member.getId());

        // then
        Reaction savedReaction = reactionRepository.findById(reaction.getId()).orElseThrow();
        assertThat(savedReaction.getStatus()).isEqualTo(ReactionStatus.ACTIVE);
    }

    @Test
    @DisplayName("좋아요를 취소할 수 있다.")
    void unlike() {
        // given
        Member member = createMember();
        Reaction reaction = new Reaction("EXAM", 1L, member.getId(), ReactionType.LIKE);
        reactionRepository.save(reaction);

        // when
        likeService.removeLike(new Like(LikeTarget.EXAM, 1L), member.getId());

        // then
        Reaction savedReaction = reactionRepository.findById(reaction.getId()).orElseThrow();
        assertThat(savedReaction.getStatus()).isEqualTo(ReactionStatus.DELETED);
    }

    @Test
    @DisplayName("좋아요를 한 상태에서만 좋아요를 취소할 수 있다.")
    void unlike_notLiked() {
        // given
        Member member = createMember();
        Reaction reaction = new Reaction("EXAM", 1L, member.getId(), ReactionType.LIKE);
        reaction.delete();
        reactionRepository.save(reaction);

        // when
        likeService.removeLike(new Like(LikeTarget.EXAM, 1L), member.getId());

        // then
        Reaction savedReaction = reactionRepository.findById(reaction.getId()).orElseThrow();
        assertThat(savedReaction.getStatus()).isEqualTo(ReactionStatus.DELETED);
    }

    private Member createMember() {
        Member member = MemberTestData.defaultMember().build();
        return memberRepository.save(member);
    }
}

File: server/src/test/kotlin/com/fluffy/reaction/domain/LikeServiceTest.kt
package com.fluffy.reaction.domain

import com.fluffy.global.exception.NotFoundException
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify

class LikeServiceTest : BehaviorSpec({

    val reactionRepository = mockk<ReactionRepository>()
    val likeService = LikeService(reactionRepository)

    Given("like") {
        val like = Like(LikeTarget.EXAM, 1L)
        val memberId = 1L
        val reaction = mockk<Reaction>()
        every { reaction.id } returns 1L
        every { reaction.activate() } returns Unit

        When("이미 좋아요가 존재하는 경우") {

            every {
                reactionRepository.findByTargetTypeAndTargetIdAndMemberIdAndType(any(), any(), any(), any())
            } returns reaction

            Then("좋아요를 활성화하고, 식별자를 반환한다.") {
                val actual = likeService.like(like, memberId)
                val expected = 1L

                verify { reaction.activate() }
                actual shouldBe expected
            }
        }

        When("좋아요가 존재하지 않는 경우") {
            every {
                reactionRepository.findByTargetTypeAndTargetIdAndMemberIdAndType(any(), any(), any(), any())
            } returns null
            every {
                reactionRepository.save(any())
            } returns reaction

            Then("좋아요를 생성하고, 식별자를 반환한다.") {
                val actual = likeService.like(like, memberId)
                val expected = 1L

                actual shouldBe expected
            }
        }
    }

    Given("removeLike") {
        val like = Like(LikeTarget.EXAM, 1L)
        val memberId = 1L
        val reaction = mockk<Reaction>()
        every { reaction.id } returns 1L
        every { reaction.delete() } returns Unit

        When("좋아요가 존재하는 경우") {
            every {
                reactionRepository.findByTargetTypeAndTargetIdAndMemberIdAndType(any(), any(), any(), any())
            } returns reaction

            Then("좋아요를 비활성화하고, 식별자를 반환한다.") {
                val actual = likeService.removeLike(like, memberId)
                val expected = 1L

                verify { reaction.delete() }
                actual shouldBe expected
            }
        }

        When("좋아요가 존재하지 않는 경우") {
            every {
                reactionRepository.findByTargetTypeAndTargetIdAndMemberIdAndType(any(), any(), any(), any())
            } returns null

            Then("예외를 발생시킨다.") {
                shouldThrow<NotFoundException> {
                    likeService.removeLike(like, memberId)
                }.message shouldBe "아직 좋아요를 하지 않았습니다. 좋아요 정보: $like"
            }
        }
    }
})
